runtime/gc_glue_java/HeadlessMixedObjectScanner.hpp

```diff
@@ -51,13 +51,14 @@ class GC_HeadlessMixedObjectScanner : public GC_ObjectScanner
 
 	/**
 	 * @param env The scanning thread environment
-	 * @param[in] objectPtr the object to be processed
-	 * @param[in] flags Scanning context flags
+	 * @param scanPtr Pointer to the start of the object
+	 * @param size The instance size
+	 * @param flags Scanning context flags
 	 */
-	MMINLINE GC_HeadlessMixedObjectScanner(MM_EnvironmentBase *env, J9Class *clazzPtr, fomrobject_t *scanPtr, uintptr_t flags)
+	MMINLINE GC_HeadlessMixedObjectScanner(MM_EnvironmentBase *env, fomrobject_t *scanPtr, uintptr_t size, uintptr_t flags)
 		: GC_ObjectScanner(env, scanPtr, 0, flags)
-		, _endPtr((fomrobject_t *)((uint8_t*)_scanPtr + env->getExtensions()->mixedObjectModel.getSizeInBytesWithoutHeader(clazzPtr)))
-		, _mapPtr(_scanPtr)
+		, _endPtr((fomrobject_t *)((uint8_t*)scanPtr + size))
+		, _mapPtr(scanPtr)
```

Comment:
I'm struggling to understand why this extra field _mapPtr is necessary and why existing _scanPtr would not be able to serve the same purpose.

As it stands now _scanPtr  will stay at the first object slot within scan map, while _mapPtr will be used to make progress and point to current object slot. Is that first object slot info used anywhere?

The only spot I see that _scanPtr is used is getBytesRemaining, which if called in a middle of a scan appear to return incorrect remaining value, since _scanPtr is not current pointer, but _mapPtr is.

Reply:
`_mapPtr` is used in `getNextSlotMap()` as a base point to advance scan pointer to the slot next non-zero map word pointed to.  At the time to request new map `_scanPtr` might be not advanced to the end of scanning interval (defined by current map word) but slot corresponds with last non-zero bit in map word. So if we remove `_mapPtr` we need to find another way to calculate slot address new map word first bit pointed to.

Reply:
Yeah, I got it, thanks. No need to change the logic, although names could bit more descriptive like:
_scanPtr -> _currentObjectSlot
_mapPtr -> _firstObjectSlotInCurrentMapWord
But there are similar comments in the code, so it's fine.